### PR TITLE
Optionally depend on usertiming-compression.

### PR DESCRIPTION
**Pull request by neurolabs**
Make compression optional by adding a configuration option
"ut_compression". When set to falsy, the usertiming plugin does not depend on
the usertiming-compression library, but sends the timings uncompressed
instead.

format for uncompressed send is json, example:

  {"mark":{"foo":1234},{"measure":{"bar":5678,"baz":432}}

### CODE REVIEW

**Comment by bluesmoon** [comment]
I think it might be better to have a single flag that turns on/off compression for UT as well as RT as it doesn't make sense to have one on without the other.  Also, have a look at how other plugins use `BOOMR.utils.pluginConfig`

BTW, in our experience with compression turned off, the beacon can take multiple TCP packets to get sent through.

**Comment by neurolabs** [comment]
Thanks for the hints, I will reiterate.

We chose to not use compression because we currently don't have many usertimings measured, and are not using JS on the server side. Reimplementing the decompression would take too long for the current gain.

**Comment by bluesmoon** [comment]
Ok, so in that case, maybe a flag is not needed.  You can feature detect the compression library.  If it exists, then compress the data, if it doesn't exist, then don't.  That way you can build boomerang without the compression library resulting in a smaller JS size in addition to not compressing.

**Comment by neurolabs** [comment]
Currently, if the compression library is not available, checking and sending beacons is deferred until pageReady. I'm not sure what behaviour to implement in order to not deviate too much from that behaviour.

If it's ok to disregard the current behaviour, the simplest thing would be to just check for the compression library on a call to getUserTiming().

What do you think?

**Comment by bluesmoon** [comment]
I think that was done because the compression library could be loaded asynchronously.  Since `getUserTiming()` is called in the `onbeacon` event, I think it's safe to disregard the current behaviour.  Anyone who wants to use compression will most likely bundle the library.

I would like @nicjansma's opinion.

**Comment by nicjansma** [comment]
Sounds good to me!

**Comment by neurolabs** [comment]
updated the PR, please review.

**Comment by neurolabs** [comment]
maybe @nicjansma should take a look at the code shamelessly copied from the usertiming-compression.js in this MR (for getting the usertimings before deciding whether compressing or not.

File: plugins/usertiming.js [hunk 1]
```diff
@@ -87,21 +93,49 @@
...
+					return e.startTime + e.duration >= options.from;
+				});
+			}
+
+			var utc = window.UserTimingCompression || BOOMR.window.UserTimingCompression;
 
-			return res;
+			if (typeof utc === "undefined") {
+				var timings, res;
+				timings = utc.compressUserTiming(entries);
```
Comment (bluesmoon) [inline plugins/usertiming.js]: shouldn't this condition be swapped?  if `utc` is `undefined`, then you can't use it in here.

Comment (neurolabs) [reply plugins/usertiming.js]: oops

**Comment by neurolabs** [comment]
What's the status here? Do you guys need anything from me?

**Comment by bluesmoon** [comment]
Sorry for the delayed response. Have been out sick. The code changes look good. Can we get a couple of tests for the code?  Then I can merge.

**Comment by neurolabs** [comment]
Can you please tell me how I can write tests that run on a version of boomerang without compression? currently, when I naively run the test suite with a minimal plugins.json, most unit tests fail.

File: plugins/usertiming.js [hunk 2]
```diff
@@ -47,6 +51,8 @@
  *     {"name":"mark2","startTime":200,"duration":0,"entryType":"mark"},
  *     {"name":"measure2","startTime":200,"duration":200,"entryType":"measure"}]
  *
+ * Uncompressed measures only send their duration in contrast to comrpressed measures, which also send their startTime.
```
Comment (nicjansma) [inline plugins/usertiming.js]: s/comrpressed/compressed/

Comment (nicjansma) [reply plugins/usertiming.js]: For your use-case, you're not using a measure's `startTime` for anything?  Are we sure other consumers won't want access to the `startTime`?

File: plugins/usertiming.js [hunk 3]
```diff
@@ -87,21 +93,49 @@
 		options: {"from": 0, "window": BOOMR.window},
 
 		/*
+		 * Gets the user timings, filters out those that have already been sent
```
Comment (nicjansma) [inline plugins/usertiming.js]: Ideally has a `.` at the end of the sentence so it doesn't run when JSDoc is compiled.

File: plugins/usertiming.js [hunk 4]
```diff
@@ -87,21 +93,49 @@
...
+		 * data that occurred since the last call
 		 *
-		 * @returns {string} Compressed UserTiming data
+		 * @returns {string} UserTiming data
 		 */
 		getUserTiming: function() {
-			var timings, res,
-			    now = BOOMR.hrNow(),  // Get a timestamp to compare with performance marks and measures.
-			    utc = window.UserTimingCompression || BOOMR.window.UserTimingCompression;
+			var frame, entries, options;
```
Comment (nicjansma) [inline plugins/usertiming.js]: I think we could simplify this (realizing it was migrated from https://github.com/nicjansma/usertiming-compression.js/blob/master/src/usertiming-compression.js#L381) to just using the `impl.options` always and not having a local `options`.

File: plugins/usertiming.js [hunk 5]
```diff
@@ -87,21 +93,49 @@
...
 
-			timings = utc.getCompressedUserTiming(impl.options);
-			res = utc.compressForUri(timings);
-			this.options.from = now;
+			options = impl.options || {};
+			frame = options.window || window;
+			entries = this.findUserTimingForFrame(frame);
```
Comment (nicjansma) [inline plugins/usertiming.js]: Then this would be just `entries = this.findUserTimingForFrame(options.window);`

**Comment by nicjansma** [comment]
@neurolabs For testing, I'm hoping some of the guides in https://github.com/SOASTA/boomerang/blob/master/doc/tests.md will make the approach a bit clearer now.

I'm thinking you could add a new test or two to the E2E `18-usertiming` test suite:
https://github.com/SOASTA/boomerang/tree/master/tests/page-templates/18-usertiming

Probably the easiest way is to use the default `plugins.json` (which includes `usertiming.js` and `node_modules/usertiming-compression/src/usertiming-compression.js`), which builds `tests/build/boomerang-latest-debug.js`.

Start with something like https://github.com/SOASTA/boomerang/blob/master/tests/page-templates/18-usertiming/01-usertiming-basic.html

Change `<%= boomerangSnippet %>` to `<%= boomerangScript %>` so Boomerang is loaded inline, instead of using the async loader pattern.

Then the next script line could be:

```
<script>
delete window.UserTimingCompression;
</script>
```

This would remove the library for that test only, and the rest of the test could basically be run as-is.

Then in the new `.js` file (based on https://github.com/SOASTA/boomerang/blob/master/tests/page-templates/18-usertiming/01-usertiming-basic.js), you'd modify the validations to look for the uncompressed JSON.

Hope that makes sense :) Let me know of any questions.

File: plugins/usertiming.js (hunk 2 above)
Comment (neurolabs) [reply plugins/usertiming.js]: If you want me to, I can integrate it, but we don't measure it, since we can interpret the corresponding marks.

**Comment by neurolabs** [comment]
Thanks, @nicjansma , your comments and your idea for testing make perfect sense. Please provide an answer to my single reply.

And keep in mind that I will be gone the next three weeks. I will finalize this.

File: plugins/usertiming.js (hunk 2 above)
Comment (nicjansma) [reply plugins/usertiming.js]: That's fine then, we can keep it at just `duration` for now and if someone else needs `startTime` we can add an option

**Comment by neurolabs** [comment]
@nicjansma added the tests and addressed your feedback. I have my fingers crossed for a direct merge ;).

**Comment by nicjansma** [comment]
Sorry for the delays, this looks great!

**Pull request merged**